<a href="https://colab.research.google.com/github/783atulpatel/Colab_REPO/blob/main/mapreduce.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**MapReduce (Basic Job Execution)**
###What We will Learn in this :
####how to run a MapReduce job, including creating and executing jobs on HDFS.




# Install Java Development Kit (JDK

In [2]:

!apt-get install openjdk-8-jdk-headless -qq > /dev/null

# setup Hadoop as told in previous class

In [4]:
!tar -xzf /content/drive/MyDrive/hadoop-setup/hadoop-3.3.6.tar.gz -C /content/

# setting up environment

In [5]:
# Set Hadoop environment variables
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["HADOOP_HOME"] = "/content/hadoop-3.3.6"
os.environ["PATH"] = os.environ["PATH"] + ":" + os.environ["HADOOP_HOME"] + "/bin" + ":" + os.environ["HADOOP_HOME"] + "/sbin"

# Configure Hadoop (replace with your desired configuration)
!echo "export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64" >> ~/.bashrc
!echo "export HADOOP_HOME=/content/hadoop-3.3.6" >> ~/.bashrc
!echo "export PATH=$PATH:$HADOOP_HOME/bin:$HADOOP_HOME/sbin" >> ~/.bashrc

# Source the bashrc file to apply the changes
!source ~/.bashrc

# Verify Hadoop installation
!hadoop version

Hadoop 3.3.6
Source code repository https://github.com/apache/hadoop.git -r 1be78238728da9266a4f88195058f08fd012bf9c
Compiled by ubuntu on 2023-06-18T08:22Z
Compiled on platform linux-x86_64
Compiled with protoc 3.7.1
From source with checksum 5652179ad55f76cb287d9c633bb53bbd
This command was run using /content/hadoop-3.3.6/share/hadoop/common/hadoop-common-3.3.6.jar


# verfying the working of haddop

In [6]:
!jps

8597 Jps


# so initially will create an input.txt file

In [22]:

with open('/content/new_folder/input.txt', 'w') as f:
    f.write("Hello HadoopHadoop MapReduce is great for big dataBig data can be processed using Hadoop MapReduce is a programming model")

In [23]:
!hadoop jar /content/hadoop-3.3.6/share/hadoop/mapreduce/hadoop-mapreduce-examples-3.3.6.jar wordcount /content/new_folder/input.txt /content/new_folder/output

2025-01-04 11:19:02,561 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2025-01-04 11:19:02,735 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2025-01-04 11:19:02,736 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2025-01-04 11:19:03,085 INFO input.FileInputFormat: Total input files to process : 1
2025-01-04 11:19:03,143 INFO mapreduce.JobSubmitter: number of splits:1
2025-01-04 11:19:03,510 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local884788319_0001
2025-01-04 11:19:03,511 INFO mapreduce.JobSubmitter: Executing with tokens: []
2025-01-04 11:19:03,755 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
2025-01-04 11:19:03,757 INFO mapreduce.Job: Running job: job_local884788319_0001
2025-01-04 11:19:03,769 INFO mapred.LocalJobRunner: OutputCommitter set in config null
2025-01-04 11:19:03,784 INFO output.PathOutputCommitterFactory: No output committer factory defined, defaultin

In [24]:
#displaying the result of MapReduce WordCount Program
!hdfs dfs -cat /content/new_folder/output/part-r-00000

Hadoop	1
HadoopHadoop	1
Hello	1
MapReduce	2
a	1
be	1
big	1
can	1
data	1
dataBig	1
for	1
great	1
is	2
model	1
processed	1
programming	1
using	1


**So here we have seen excution of MapReduce using Existing jar Files**

# Working On New dataset using own Scripting Of map Reduce

**Case Study 2: Titanic Dataset - Survival Analysis**
**Dataset Source:**
Titanic Dataset

**Task:**
Count the number of survivors and non-survivors.

**Steps:**
Prepare the Data: Download and upload



1. the train.csv file to /content/newfolder.

2. Write the MRJob Script: Save this as survival_analysis.py.

### downloading the dataset

In [26]:


!wget https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv -O /content/new_folder/train.csv


--2025-01-04 11:29:36--  https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60302 (59K) [text/plain]
Saving to: ‘/content/new_folder/train.csv’

/content/new_folder 100%[===================>]  58.89K  --.-KB/s    in 0.004s  

2025-01-04 11:29:36 (12.9 MB/s) - ‘/content/new_folder/train.csv’ saved [60302/60302]



### viewing the dataset

In [27]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


df= pd.read_csv('/content/new_folder/train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### writing the script to count the passenger

In [25]:
%%writefile survival_analysis.py
from mrjob.job import MRJob

class SurvivalAnalysis(MRJob):

    def mapper(self, _, line):
        fields = line.split(',')
        if fields[0] != "PassengerId":  # Skip the header
            survived = fields[1]
            yield survived, 1

    def reducer(self, survived, counts):
        yield survived, sum(counts)

if __name__ == "__main__":
    SurvivalAnalysis.run()


Writing survival_analysis.py


# Run the Script:

In [28]:
!python3 survival_analysis.py /content/new_folder/train.csv > /content/new_folder/survival_analysis_output.txt


No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory /tmp/survival_analysis.root.20250104.113137.228419
Running step 1 of 1...
job output is in /tmp/survival_analysis.root.20250104.113137.228419/output
Streaming final output from /tmp/survival_analysis.root.20250104.113137.228419/output...
Removing temp directory /tmp/survival_analysis.root.20250104.113137.228419...


#View the Results:

In [30]:
!cat /content/new_folder/survival_analysis_output.txt


"0"	549
"1"	342


here we have found that 342 paasenger survived where as 549 not

In [32]:
!jps

48618 Jps


In [34]:
# prompt: update the below survival analysis .py code to  count total no of passenger  and no of passeneger male and female who died using the dataset  using mrjob
%%writefile TitanicSurvivalAnalysis.py
from mrjob.job import MRJob
from mrjob.step import MRStep

class TitanicSurvivalAnalysis(MRJob):
    def steps(self):
        return [
            MRStep(mapper=self.mapper_get_survival_info,
                   reducer=self.reducer_count_survival)
        ]

    def mapper_get_survival_info(self, _, line):
        try:
            passenger_id, survived, pclass, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked = line.split(',')
            if passenger_id == 'PassengerId':  # Skip header row
                return
            yield survived, (sex, 1) #Yield survived status with sex and a count
        except ValueError:
            pass #Handle lines that don't have the expected format


    def reducer_count_survival(self, survived, passenger_info):
        total_survivors = 0
        male_deaths = 0
        female_deaths = 0
        total_passengers = 0
        for sex, count in passenger_info:
            total_passengers += count
            if survived == '0':
                if sex == 'male':
                    male_deaths += count
                elif sex == 'female':
                    female_deaths += count
        total_survivors = total_passengers # Initialize total survivors
        if survived == '0':
            total_survivors = 0 # reset if it's a death count
        yield survived, (total_passengers, male_deaths, female_deaths)

if __name__ == '__main__':
    TitanicSurvivalAnalysis.run()

Writing TitanicSurvivalAnalysis.py


In [35]:
!python3 survival_analysis.py /content/new_folder/train.csv > /content/new_folder/survival_analysis_output.txt

No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory /tmp/survival_analysis.root.20250104.123216.311927
Running step 1 of 1...
job output is in /tmp/survival_analysis.root.20250104.123216.311927/output
Streaming final output from /tmp/survival_analysis.root.20250104.123216.311927/output...
Removing temp directory /tmp/survival_analysis.root.20250104.123216.311927...


In [36]:
!cat /content/new_folder/survival_analysis_output.txt

"0"	549
"1"	342
